src/components/Modal/RunJobModal.js

```diff
@@ -1,30 +1,39 @@
-import React from 'react'
+import React, { useState } from 'react'
 import { PropTypes } from '@dhis2/prop-types'
-import { useDataEngine } from '@dhis2/app-runtime'
+import { useDataMutation } from '@dhis2/app-runtime'
 import {
     Button,
     Modal,
     ModalContent,
     ModalActions,
     ButtonStrip,
+    NoticeBox,
 } from '@dhis2/ui'
 import i18n from '@dhis2/d2-i18n'
 import { hooks } from '../Store'
 
 const RunJobModal = ({ id, hideModal }) => {
-    const engine = useDataEngine()
-    const query = {
-        jobs: {
-            resource: `jobConfigurations/${id}/execute`,
+    const [mutation] = useState(() => ({
+        resource: `jobConfigurations/${id}/execute`,
+        type: 'create',
+    }))
```

Comment:
So I assume that this works because react's useState allows passing a function to setState, which will receive the previous state and has to return the new state. To me the semantics of this are not very clear. Essentially this is doing the same as memoizing the mutation right? Or is there another reason that makes you prefer this approach.

I really think we should settle on one recommended approach for this pattern as an org. btw. This is I think the third pattern I've seen us use and that could get difficult to maintain further down the line.

Reply:
> So I assume that this works because react's useState allows passing a function to setState, which will receive the previous state and has to return the new state

No this is not related to functional state updates - passing a function to `useState` allows calculating initial value of `mutation` just once. This is similar to `useMemo` except that [`useMemo` is not a semantic guarantee, it's just a performance optimisation](https://reactjs.org/docs/hooks-reference.html#usememo).

Reply:
> I really think we should settle on one recommended approach for this pattern as an org. btw. This is I think the third pattern I've seen us use and that could get difficult to maintain further down the line.

The recommended approach for constants inside a functional component, once we have a DHIS2 hooks lib, will probably be [this `useConst` hook](https://gist.github.com/amcgee/42bb2fa6d5f79e607f00e6dccc733482).

Reply:
> No this is not related to functional state updates - passing a function to useState allows calculating initial value of mutation just once.

So why not just pass the desired object to initial state directly? That would also set it only once and skip using a callback. Right?

Reply:
In other words, it is my understanding that initial state will only be set once: during state initialization (which makes sense, given the name initial state), and persist as long as the component is mounted. Whether you use a callback or directly pass the desired initial value shouldn't matter. That is my understanding, or am I overlooking something?

Reply:
Yep you're right it shouldn't matter whether it's a callback or passed directly, I think I just did it this way since I started with `useMemo` then remembered that it's not the 'correct' solution. I'll remove the callback :+1:

Reply:
Thanks. Yeah I was trying to understand whether I missed something since it confused me a little. Tbh, I think I prefer this solution (or useRef if the setting shouldn't trigger a rerender) over a useConst hook. Especially over maintaining our own react hook lib. But I digress.

> This is similar to useMemo except that useMemo is not a semantic guarantee, it's just a performance optimisation.

Yeah you're right, I forgot about that one. That's a good reason. Thanks for elaborating!